pkg/cloudprovider/provider/azure/provider.go

```diff
@@ -0,0 +1,581 @@
+package azure
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-04-01/compute"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/golang/glog"
+
+	"github.com/kubermatic/machine-controller/pkg/cloudprovider/cloud"
+	"github.com/kubermatic/machine-controller/pkg/cloudprovider/common/ssh"
+	cloudprovidererrors "github.com/kubermatic/machine-controller/pkg/cloudprovider/errors"
+	"github.com/kubermatic/machine-controller/pkg/cloudprovider/instance"
+	"github.com/kubermatic/machine-controller/pkg/machines/v1alpha1"
+	"github.com/kubermatic/machine-controller/pkg/providerconfig"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	machineUIDTag = "Machine-UID"
+	adminUserName = "kubermatic"
+)
+
+type provider struct {
+	configVarResolver *providerconfig.ConfigVarResolver
+}
+
+type rawConfig struct {
+	SubscriptionID providerconfig.ConfigVarString `json:"subscriptionID"`
+	TenantID       providerconfig.ConfigVarString `json:"tenantID"`
+	ClientID       providerconfig.ConfigVarString `json:"clientID"`
+	ClientSecret   providerconfig.ConfigVarString `json:"clientSecret"`
+
+	Location      providerconfig.ConfigVarString `json:"location"`
+	ResourceGroup providerconfig.ConfigVarString `json:"resourceGroup"`
+	VMSize        providerconfig.ConfigVarString `json:"vmSize"`
+	VNetName      providerconfig.ConfigVarString `json:"vnetName"`
+	SubnetName    providerconfig.ConfigVarString `json:"subnetName"`
+}
+
+type config struct {
+	SubscriptionID string
+	TenantID       string
+	ClientID       string
+	ClientSecret   string
+
+	Location      string
+	ResourceGroup string
+	VMSize        string
+	VNetName      string
+	SubnetName    string
+}
+
+type azureVM struct {
+	vm          *compute.VirtualMachine
+	ipAddresses []string
+	status      instance.Status
+}
+
+func (vm *azureVM) Addresses() []string {
+	return vm.ipAddresses
+}
+
+func (p *azureVM) ID() string {
+	return *p.vm.ID
+}
+
+func (p *azureVM) Name() string {
+	return *p.vm.Name
+}
+
+func (p *azureVM) Status() instance.Status {
+	return p.status
+}
+
+var imageReferences = map[providerconfig.OperatingSystem]compute.ImageReference{
+	providerconfig.OperatingSystemCoreos: compute.ImageReference{
+		Publisher: to.StringPtr("CoreOS"),
+		Offer:     to.StringPtr("CoreOS"),
+		Sku:       to.StringPtr("Stable"),
+		Version:   to.StringPtr("latest"),
+	},
+	providerconfig.OperatingSystemCentOS: compute.ImageReference{
+		Publisher: to.StringPtr("OpenLogic"),
+		Offer:     to.StringPtr("CentOS"),
+		Sku:       to.StringPtr("7.4"),
```

Comment:
The current CentOS 7.5 is not available on Azure?

Reply:
Cannot find it. Also, there doesn't seem to be an "official" image.